### PR TITLE
Add Token and OAuth2 authentication methods

### PR DESCRIPTION
**Pull request by decko**
Review Checklist:
- [ ] An issue is properly linked. [feature and bugfix only]
- [ ] Tests are present or not feasible.
- [ ] Commits are split in a logical way (not historically).

### CODE REVIEW

File: pulp-glue/pulp_glue/common/openapi.py [hunk 1]
```diff
@@ -53,36 +55,79 @@ class AuthProviderBase:
     Returned auth objects need to be compatible with `requests.auth.AuthBase`.
     """
 
-    def basic_auth(self) -> t.Optional[t.Union[t.Tuple[str, str], requests.auth.AuthBase]]:
```
Comment (mdellweg) [inline pulp-glue/pulp_glue/common/openapi.py:56]: No, you don't get to change the interface.

File: pulp-glue/pulp_glue/common/openapi.py [hunk 2]
```diff
@@ -53,36 +55,79 @@ class AuthProviderBase:
...
-    ) -> t.Optional[t.Union[t.Tuple[str, str], requests.auth.AuthBase]]:
-        for proposal in security:
-            if [security_schemes[name] for name in proposal] == [
-                {"type": "http", "scheme": "basic"}
-            ]:
-                result = self.basic_auth()
-                if result:
-                    return result
-        raise OpenAPIError(_("No suitable auth scheme found."))
+    def __call__(self, request: requests.PreparedRequest) -> requests.PreparedRequest:
```
Comment (mdellweg) [inline pulp-glue/pulp_glue/common/openapi.py:64]: This is an Auth provider. It's supposed to provide appropriate AuthBase objects. It's not authenticating anything.

File: pulp-glue/pyproject.toml [hunk 3]
```diff
@@ -25,6 +25,7 @@ classifiers = [
 dependencies = [
   "packaging>=20.0,<24",
   "requests>=2.24.0,<2.32",
+  "requests-oauthlib>=1.4.0,<1.5.0"
```
Comment (dkliban) [inline pulp-glue/pyproject.toml:28]: you are missing a comma at the end.

File: pulp-glue/pulp_glue/common/openapi.py [hunk 4]
```diff
@@ -53,36 +55,79 @@ class AuthProviderBase:
...
+        self.token = token
+
+    def auth(self, request) -> requests.PreparedRequest:
+        request.headers["Authorization"] = self.token
+        return request
+
+
+class ThirdPartyAuthProvider(AuthProviderBase):
+
+    auth_type = "third_party"
```
Comment (mdellweg) [inline pulp-glue/pulp_glue/common/openapi.py:114]: Why can we not just call it oauth2?
Also isn't there a requests-oauth out there?

File: pulp_cli/config.py [hunk 5]
```diff
@@ -58,6 +61,9 @@
     ),
     click.option("--username", default=None, help=_("Username on pulp server")),
     click.option("--password", default=None, help=_("Password on pulp server")),
+    click.option("--client_id", default=None, help=_("client_id on Identity Provider")),
```
Comment (mdellweg) [inline pulp_cli/config.py:64]: Username?

File: pulp_cli/config.py [hunk 6]
```diff
@@ -58,6 +61,9 @@
...
+    click.option("--client_secret", default=None, help=_("client_secret on Identity Provider")),
```
Comment (mdellweg) [inline pulp_cli/config.py:65]: password?

File: pulp_cli/config.py [hunk 7]
```diff
@@ -58,6 +61,9 @@
...
+    click.option("--token_url", default=None, help=_("URL to reach the Identity Provider")),
```
Comment (mdellweg) [inline pulp_cli/config.py:66]: The user does not know that. It must be extracted from the www-authenticate header.

**Review by mdellweg — CHANGES_REQUESTED** [review]
This is heading in the wrong direction.

File: pulp-glue/pulp_glue/common/openapi.py (hunk 4 above)
Comment (decko) [reply pulp-glue/pulp_glue/common/openapi.py]: Renamed it, and yes, we're using `request-oauthlib`. Here https://github.com/pulp/pulp-cli/blob/0738418f7247b7e901c03ea69b6997cb6842653e/pulp-glue/pyproject.toml#L28

File: pulp-glue/pulp_glue/common/openapi.py (hunk 2 above)
Comment (decko) [reply pulp-glue/pulp_glue/common/openapi.py:64]: It's returning an `AuthBase` compatible object, like stated [here](https://github.com/pulp/pulp-cli/blob/375f8b0952e33c409eed58c396fe0feedb7e43eb/pulp-glue/pulp_glue/common/openapi.py#L55) and based on [this](https://requests.readthedocs.io/en/latest/_modules/requests/auth/#AuthBase).
If this is about the name of the `auth` method it was based on the previous one `basic_auth`, which I think it's related only to HTTP Basic Authentication method.

File: pulp-glue/pulp_glue/common/openapi.py (hunk 1 above)
Comment (decko) [reply pulp-glue/pulp_glue/common/openapi.py:56]: Yeah, but the way the `AuthProviderBase` is written the `__call__` method is incompatible with the original interface and also only calls the `basic_auth` method. This way every time we need to write a new AuthProvider we have to override the `__call__` method to: first, be compliant with `AuthBase` objects, and second: to call another method that would make more send for different authentication method.

File: pulp-glue/pulp_glue/common/openapi.py [hunk 8]
```diff
@@ -133,6 +178,7 @@ def __init__(
             if cert or key:
                 raise OpenAPIError(_("Cannot use both 'auth' and 'cert'."))
             self.auth_provider = auth_provider
+            self._session.auth = auth_provider
```
Comment (decko) [inline pulp-glue/pulp_glue/common/openapi.py:181]: need to check for a better way to provide the authentication if needed.

File: pulp_cli/config.py (hunk 7 above)
Comment (decko) [reply pulp_cli/config.py:67]: Although the OAuth2 RFC presents a `WWW-Authenticate` section [here](https://www.rfc-editor.org/rfc/rfc6750#section-3), it's not clear that it should contain the issuer URL where we could ask for a new token, for example. 
Also, I could be wrong here.

File: pulp-glue/pulp_glue/common/openapi.py (hunk 2 above)
Comment (mdellweg) [reply pulp-glue/pulp_glue/common/openapi.py:64]: It is supposed to _return_, not _be_ an AuthBase.

**Comment by decko** [comment]
> This is heading in the wrong direction.

Ok, starting over on another PR.

**Comment by mdellweg** [comment]
> > This is heading in the wrong direction.
> 
> Ok, starting over on another PR.

Please don't. That way we loose all the discussion. Please drop the new commits in the current PR.

**Pull request closed without merging**